### DISCUSSION TITLE
Connection List

### DISCUSSION THREAD

**Discussion by go4cas** [pocketbase/pocketbase#4166] [Q&A] [answered]
Noob question: I was trying to keep track of all realtime connections, by pushing to a global array variable from `onRealtimeConnectRequest`. Then have a separate GET route that would list the connections array as a json payload.

I have realised, from the docs, that this global variable scope is not possible.

Is the only other way to achieve my use, by creating a `connections` collection, and then add/remove entries as clients connect/disconnect? Or, are there other options?

**Comment by ganigeorgiev** [comment]
If you just want a list/map with the current realtime connected clients that you can return in your custom GET route, then you can use [`$app.subscriptionsBroker().clients()`](https://pocketbase.io/jsvm/interfaces/subscriptions.Broker.html#clients).

If you are trying to implement realtime "presence"-like functionality then you can explore some of the suggested workarounds in https://github.com/pocketbase/pocketbase/discussions/3378.

**Reply by go4cas** [reply]
Thanks, @ganigeorgiev! That will help!

When calling `clientById()`, I get no result, although I do get the id back in the `clients()` call.

```javascript
routerAdd('GET', '/connections', (c) => {
  const sb = $app.subscriptionsBroker()
  const clients = sb.clients()
  console.log('CLIENTS:', JSON.stringify(clients))
  const clientList = Object.keys(clients).map((key) => {
    console.log('KEY:', key)
    const client = sb.clientById(key)
    console.log('C:', JSON.stringify(client))
    return client
  })
  console.log('CONN:', JSON.stringify(clientList))
  return c.json(200, { clientList })
})
```
<img width="550" alt="Screenshot 2024-01-17 at 19 15 58" src="https://github.com/pocketbase/pocketbase/assets/12674784/fe75076c-124d-4f37-a623-6e90cea3937c">

**Reply by ganigeorgiev** [reply, marked as answer]
Most of the time the JSVM operates with Go translated objects, aka. non-plain JS objects, meaning that some objects may not have public fields to show (to confirm that the client is not empty you can call `client.id()`; for all methods check the [Client interface docs](https://pocketbase.io/jsvm/interfaces/subscriptions.Client.html)).

I'm not sure that I understand what is the desired response you want to return from your route, but note that it may not be very efficient to call `sb.clientById` because you already have a shallow copy of all clients. A better approach would be to iterate directly over the fetched map.

**Reply by go4cas** [reply]
Okay, I see the `client` only contains methods. So, if I do `const client = sb.clientById(key).id()` it seems to bring back data.
